group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -428,9 +432,51 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        List<CompletableFuture<ListGroupsResponseData>> futures = new ArrayList<>();
+        for (TopicPartition tp : runtime.partitions()) {
+            futures.add(runtime.scheduleReadOperation(
+                "list-groups",
+                tp,
+                (coordinator, lastCommittedOffset) -> coordinator.listGroups(request.statesFilter(), lastCommittedOffset)
+            ).exceptionally(exception -> {
+                if (!(exception instanceof KafkaException)) {
+                    log.error("ListGroups request {} hit an unexpected exception: {}",
+                        request, exception.getMessage());
+                    throw new RuntimeException(exception);
+                }
+                if (exception instanceof CoordinatorLoadInProgressException) {
+                    throw new RuntimeException(exception);
+                } else if (exception instanceof NotCoordinatorException) {
+                    log.warn("ListGroups request {} hit a NotCoordinatorException exception: {}",
+                        request, exception.getMessage());
+                    return new ListGroupsResponseData().setGroups(Collections.emptyList());
+                } else {
+                    return new ListGroupsResponseData().setErrorCode(Errors.forException(exception).code());
+                }
+            }));
+        }
+        CompletableFuture<ListGroupsResponseData> responseFuture = new CompletableFuture<>();
+        List<ListGroupsResponseData.ListedGroup> listedGroups = new ArrayList<>();
+        AtomicInteger succeedFutureCount = new AtomicInteger();
+        FutureUtils.drainFutures(futures, (data, t) -> {
+            synchronized (runtime) {
+                if (t != null) {
+                    responseFuture.completeExceptionally(new UnknownServerException(t.getMessage()));
+                } else {
+                    if (data.errorCode() != Errors.NONE.code()) {
+                        if (!responseFuture.isDone()) {
+                            responseFuture.complete(data);
+                        }
+                    } else {
+                        listedGroups.addAll(data.groups());
+                        if (succeedFutureCount.addAndGet(1) == runtime.partitions().size()) {
+                            responseFuture.complete(new ListGroupsResponseData().setGroups(listedGroups));
+                        }
+                    }
+                }
+            }
+        });
+        return responseFuture;
```

Comment:
There are a few issues with this code.
1. Synchronising on `runtime` will create lock contention across all the callers of `listGroups`. We should rather use a local variable.
2. The error handling seems error prone to me. For instance, `NotCoordinatorException` exceptions are turned into `RuntimeException` exceptions and then turned into `UnknownServerException` if I understood it correctly. We lose the semantic along the way.

I think that we could take your idea further and combine the two main steps into one. I am thinking about something like this:

```
        final List<TopicPartition> partitions = new ArrayList<>(runtime.partitions());
        final CompletableFuture<ListGroupsResponseData> future = new CompletableFuture<>();
        final List<ListGroupsResponseData.ListedGroup> results = new ArrayList<>();
        final AtomicInteger cnt = new AtomicInteger(partitions.size());

        for (TopicPartition partition : partitions) {
            runtime.scheduleReadOperation(
                "list-group",
                partition,
                (coordinator, lastCommittedOffset) -> coordinator.listGroups(request.statesFilter(), lastCommittedOffset)
            ).handle((groups, exception) -> {
                if (exception == null) {
                    synchronized (results) {
                        results.addAll(groups);
                    }
                } else {
                    if (!(exception instanceof NotCoordinatorException)) {
                        future.complete(new ListGroupsResponseData().setErrorCode(Errors.forException(exception).code()));
                    }
                }

                if (cnt.decrementAndGet() == 0 && !future.isDone()) {
                    future.complete(new ListGroupsResponseData().setGroups(results));
                }

                return null;
            });
        }
```

What do you think?

Note that I have not tested this code at all. I just wrote it here.

Reply:
Thank you for your suggestions. The code is much simpler and more elegant.